vcr/stubs/httpx_stubs.py

```diff
@@ -27,17 +27,22 @@ def _transform_headers(httpx_response):
     out = {}
     for key, var in httpx_response.headers.raw:
         decoded_key = key.decode("utf-8")
+
+        # We drop the Content-Encoding header as the content
+        # we store in the response will already have been
+        # processed.
+        if decoded_key.lower() == "content-encoding":
+            continue
         out.setdefault(decoded_key, [])
         out[decoded_key].append(var.decode("utf-8"))
     return out
 
 
-def _to_serialized_response(httpx_response):
+def _to_serialized_response(resp):
     return {
-        "status_code": httpx_response.status_code,
-        "http_version": httpx_response.http_version,
-        "headers": _transform_headers(httpx_response),
-        "content": httpx_response.content.decode("utf-8", "ignore"),
+        "status": dict(code=resp.status_code, message=resp.reason_phrase),
+        "headers": _transform_headers(resp),
+        "body": {"string": resp.content},
```

Comment:
Question -- does this really generate the body content as a yaml binary string as in your[test example](https://github.com/kevin1024/vcrpy/pull/649/files#diff-e9c5392f3e558174781dc10a0f66b79455ef38846fdef971c9bb95cd70dd7d19) above? I believe that is just for reading back a cassette that has this type of content, but a new cassette generated with this stub will not be encoded the same way the others will be (or at least with requests library).

This is effectively what I needed to be able to generate the body content in the same yaml binary format:

```
from ..serializers.yamlserializer import serialize
...

def _to_serialized_response(resp):
    return {
        "status": dict(code=resp.status_code, message=resp.reason_phrase),
        "headers": _transform_headers(resp),
        "body": {"string": serialize(resp.content).encode("unicode_escape")},
    }
```

Can you confirm? And if not outright add this into PR, can we look at having an option we can set so this will be possible?

Reply:
Perhaps I misunderstand.

Just to compare requests and httpx, I ran this in a Python prompt based on my branch:
```
with vcr.use_cassette('httpx_and_requests.yaml'):
    requests.get('https://httpbin.org/json')
    httpx.get('https://httpbin.org/json')
    requests.get('https://httpbin.org/gzip')
    httpx.get('https://httpbin.org/gzip')
```

Using that, I get [this output generated](https://gist.github.com/the-allanc/fcb9ea8c260050211cb198b198b1798a) - for the JSON resource, both httpx and requests generate a regular string (without the `!!binary` marker). For the GZIP resource, requests creates a binary string, but httpx does not.

I'm able to load the requests generated by the httpx stub with the requests library, and vice versa. If I retrieve binary content (like a JPEG image) using httpx, and capture that in a cassette, then I get the `!!binary` marker.

The only reason that the gzip resource differs between requests and httpx is because requests seems to allow you to get the raw data stream (which hasn't been decoded - i.e. un-gzipped), whereas I had trouble accomplishing that with httpx - so we store the decoded (un-gzipped) text and remove the Content-Encoding response header.

Reply:
I get your test, and I don't think you're misunderstanding anything.

Attempting to make httpx _not_ decompress is its own can of worms and isn't possible without modifying source. The upshot is that we can't generate cassettes with httpx that have binary encoded yaml like we did with requests -- and we considered that a feature, for various reasons.

So at least allowing the option on whether to always store content as binary encoded would be greatly appreciated!

Reply:
@hans2520 OK, I've spent more time on this one...

After stepping into the httpx source code, I figured out that if you stream the response, and then manipulate the headers before httpx tries to process it, then you can get access to the raw data stream. So that's what I did.

I've made changes to try to prevent httpx reading a response, then reading the content without decoding, and then finally amending the response to still make it usable to regular clients that will want the content decoded normally.

httpx does still automatically read the content of responses on some occasions (particularly on redirects), so we still have to support the serialisation of decoded responses - so that code still exists in that form. I've added a new "client" class for _test_httpx_ which will automatically read responses to show that we handle decoded responses just as well as their non-decoded equivalents.

Hope you find this set of changes much more to your liking... :smile:

Reply:
Yes indeed, you are correct about manipulating the httpx headers (I don't think streaming was a requirement actually, but you would have needed to patch two different locations in my test... which was over a week ago, sorry can't remember exactly. But for the streaming -- if it works, why not).

I haven't had a chance your modification just yet, but just glancing at the commit you made and inferring from your comment, the only thing I'm wondering about is if the format of the response will actually be "raw" content (i.e. unicode character sequences) vs the yaml binary format. That's what happened in my testing when I modified the httpx source so as to not decode the response, and why I ultimately preferred not doing that but making the suggestion above to use the yaml serializer to get yaml binary format. 

If that's true, the one nice thing about yaml binary format is plays with margins nicely. But the main reason we were wanting it is so that in the end we would not have to regenerate a bunch of cassette formats we had in source control, theoretically if only the client changed but the response stayed the same (and it was still yaml binary encoded), none of that responses would change...

Reply:
The format of the cassettes that are generated for httpx seem to be the same as the ones generated by requests (at least with my latest set of changes).